File: 알고리즘/순열의역전쌍개수.py
import sys

input = sys.stdin.readline

answer = 0


def mearge(arr, start, mid, end):
    global answer
    print(start, mid, end)

    a = start
    b = mid + 1
    t = 0
    temp = [0 for _ in range(end - start + 1)]
    # print(f"temp: {temp}")
    while a <= mid and b <= end:
        if arr[a] <= arr[b]:
            temp[t] = arr[a]
            a += 1
            t += 1
        else:
            temp[t] = arr[b]
            b += 1
            t += 1
            answer += mid + 1 - a
        # print(arr)

    while a <= mid:
        temp[t] = arr[a]
        a += 1
        t += 1
    while b <= end:
        temp[t] = arr[b]
        b += 1
        t += 1
    # print(temp)

    for i in range(end - start + 1):
        arr[start + i] = temp[i]
    # print(f"arr: {arr}, answer: {answer}")
    return temp


def devide(arr, start, end):
    mid = (start + end) // 2
    if end - start == 0:
        return
    devide(arr, start, mid)
    devide(arr, mid + 1, end)
    mearge(arr, start, mid, end)
    return


t = int(input())

for _ in range(t):
    answer = 0
    n = int(input())
    arr = list(map(int, input().split(" ")))
    devide(arr, 0, len(arr) - 1)
    print(answer)

File: 알고리즘/순열의역전쌍개수.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

int answer = 0;
void mearge(vector<int> &arr, int start, int mid, int end)
{
    int a = start;
    int b = mid + 1;
    int t = 0;

    // cout << "start: " << start << ", mid: " << mid << ", end: " << end << endl;

    vector<int> temp(end - start + 1);

    while (a <= mid && b <= end)
    {
        if (arr[a] <= arr[b])
        {
            temp[t] = arr[a];
            a++;
            t++;
        }
        else
        {
            temp[t] = arr[b];
            b += 1;
            t += 1;
            answer = answer + mid + 1 - a;
        }
    }
    while (a <= mid)
    {
        temp[t] = arr[a];
        a++;
        t++;
    }
    while (b <= end)
    {
        temp[t] = arr[b];
        b++;
        t++;
    }

    for (int p = 0; p < temp.size(); p++)
    {
        arr[start + p] = temp[p];
    }
    // cout << "answer: " << answer << endl;
    // for (int v : arr)
    //     cout << v << " ";
    // cout << endl;

    return;
}

void calc(vector<int> &arr, int start, int end)
{
    int mid = (start + end) / 2;
    if (end - start == 0)
    {
        return;
    }

    calc(arr, start, mid);
    calc(arr, mid + 1, end);
    mearge(arr, start, mid, end);
    return;
}

int main(void)
{
    int t;
    cin >> t;

    for (int i = 0; i < t; i++)
    {
        answer = 0;
        int n;
        cin >> n;
        vector<int> arr(n);
        for (int j = 0; j < n; j++)
        {
            cin >> arr[j];
        }

        calc(arr, 0, n - 1);

        cout << answer << endl;
    }

    return 0;
}